### PR TITLE
Add support for relational expressions with ordered types

### LINKED ISSUE #28352: Relational expressions are not supported with all ordered types
**Description:**
The spec says for relational expressions https://ballerina.io/spec/lang/master/#section_6.27

> A relational-expr tests the relative order of two values. There must be an ordered type to which the static type of both operands belong.

But we currently seem to allow them only for numeric values.

**Steps to reproduce:**
The following aren't allowed atm.
```ballerina
import ballerina/io;

public function main() {
    string a = "a";
    string b = "b";
    io:println(a >= b);

    boolean c = true;
    boolean d = false;
    io:println(c < d);
}
```

**Affected Versions:**
slalpha1

### LINKED ISSUE #29160: Relational expressions are allowed with operands of different basic types
**Description:**
The spec says for relational expressions
> A relational-expr tests the relative order of two values. There must be an ordered type to which the static type of both operands belong.

**Steps to reproduce:**
```ballerina
public function main() {
    int i = 1;
    decimal d = 1.0;

    // boolean b1 = i == d; // error: operator '==' not defined for 'int' and 'decimal'
    boolean b2 = i <= d; // No error!
}
```

**Affected Versions:**
slalpha2

### PR DESCRIPTION
**Pull request by lasinicl**
Fixes #28352
Fixes #29160

## Approach
A relational-expr tests the relative order of two values. **There must be an ordered type to which the static type of both operands belong.**

`type OrderedType ()|boolean|int|float|decimal|string|OrderedType[];` (https://github.com/ballerina-platform/ballerina-spec/blob/master/lang/lib/array.bal#L158)

This PR **breaks/disallows** checking the relative order of numeric values when the static types of the operands belong to different ordered types(which was supported until alpha3) 
ex: 
```
    int i = 1;
    decimal d = 1.0;
    boolean b2 = i <= d; // This gives a compilation error
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 1]
```diff
@@ -3004,6 +3007,246 @@ private static boolean containsType(Set<Object> valueSpace, Type type) {
...
+    /**
+     * Check if left hand side float value is greater than or equal to the right hand side float value.
+     *
+     * @param lhsValue The value on the left hand side
+     * @param rhsValue The value on the right hand side
+     * @return True if left hand side float value is greater than or equal to the right hand side float value, else
+     * false.
+     */
+    public static boolean compareValueGreaterThanOrEqual(double lhsValue, double rhsValue) {
+        return compareValues(lhsValue, rhsValue, "") >= 0;
```
Comment (grainier) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:3096]: without passing `""` in several places, you can have an overloaded method that takes in only `lhsValue` & `rhsValue`.

**Review by grainier — DISMISSED** [review]

**Comment by grainier** [comment]
LGTM. However, since there are runtime changes, shall we request a review from a runtime team member as well?

**Review by grainier — APPROVED** [review]

**Comment by lasinicl** [comment]
> LGTM. However, since there are runtime changes, shall we request a review from a runtime team member as well?

@warunalakshitha can you please review the runtime changes?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 2]
```diff
@@ -3008,6 +3011,254 @@ private static boolean containsType(Set<Object> valueSpace, Type type) {
...
+                throw ErrorUtils.createUnorderedTypesError(TYPE_NULL, rhsValue);
+            }
+            if (isAscending) {
+                return 1;
+            }
+            return -1;
+        }
+        if (rhsValue == null) {
+            if (inRelationalExpr) {
+                throw ErrorUtils.createUnorderedTypesError(lhsValue, TYPE_NULL);
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:3135]: There seem to be several Codecov comments indicating lines that aren't covered in tests. Shall we validate and add tests covering them?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java [hunk 3]
```diff
@@ -678,17 +691,16 @@ private void generateBinaryCompareIns(BIRNonTerminator.BinaryOp binaryIns, int o
...
+                return "compareValueGreaterThan";
             case IFGE:
-                return "checkDecimalGreaterThanOrEqual";
+                return "compareValueGreaterThanOrEqual";
             case IFLT:
-                return "checkDecimalLessThan";
+                return "compareValueLessThan";
             case IFLE:
-                return "checkDecimalLessThanOrEqual";
+                return "compareValueLessThanOrEqual";
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java:703]: We can remove `checkDecimalGreaterThan`, `checkDecimalGreaterThanOrEqual`, `checkDecimalLessThan` and `checkDecimalLessThanOrEqual` functions from the typechecker. right?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java [hunk 4]
```diff
@@ -657,13 +657,26 @@ private void generateBinaryCompareIns(BIRNonTerminator.BinaryOp binaryIns, int o
...
+            }
         } else if (lhsOpType.tag == TypeTags.FLOAT && rhsOpType.tag == TypeTags.FLOAT) {
-            this.mv.visitInsn(DCMPL);
-            this.mv.visitJumpInsn(opcode, label1);
-        } else if (lhsOpType.tag == TypeTags.DECIMAL && rhsOpType.tag == TypeTags.DECIMAL) {
-            String compareFuncName = this.getDecimalCompareFuncName(opcode);
+            String compareFuncName = this.getCompareFuncName(opcode);
             this.mv.visitMethodInsn(INVOKESTATIC, TYPE_CHECKER, compareFuncName,
-                    String.format("(L%s;L%s;)Z", DECIMAL_VALUE, DECIMAL_VALUE), false);
+                    String.format("(DD)Z"), false);
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java:673]:
```suggestion
                    "(DD)Z", false);
```

**Comment by chiranSachintha** [comment]
the spec says,
`A relational-expr tests the relative order of two values. There must be an ordered type to which the static type of both operands belong. The static type of the result is boolean. `
So the following program should be valid. shouldn't it?
``` ballerina
public function main() {
    float[]|string[] a = [1, 2];
    float[] b = [1, 2];
    boolean c = a > b;
}
```

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 5]
```diff
@@ -3008,6 +2976,251 @@ private static boolean containsType(Set<Object> valueSpace, Type type) {
         return true;
     }
 
+    /**
+q     * Check if left hand side ordered type value is less than right hand side ordered type value.
```
Comment (kalaiyarasiganeshalingam) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:2980]:
```suggestion
     * Check if left hand side ordered type value is less than right hand side ordered type value.
```

**Review by mohanvive — APPROVED** [review]
LGTM, Hope runtime changes are reviewed.

**Comment by lasinicl** [comment]
> LGTM, Hope runtime changes are reviewed.

Asked @warunalakshitha to review the runtime changes.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 6]
```diff
@@ -3008,6 +2976,251 @@ private static boolean containsType(Set<Object> valueSpace, Type type) {
...
     }
 
+    /**
+     * Check if left hand side ordered type value is less than right hand side ordered type value.
+     *
+     * @param lhsValue The value on the left hand side
+     * @param rhsValue The value on the right hand side
+     * @return True if left hand side ordered type value is less than right hand side ordered type value, else false.
+     */
+    public static boolean compareValueLessThan(Object lhsValue, Object rhsValue) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:2986]: Shall we move these value compare methods to a different class? IMO these are not part of the Type checker

**Review by chiranSachintha — APPROVED** [review]
LGTM.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 7]
```diff
@@ -3008,6 +2976,251 @@ private static boolean containsType(Set<Object> valueSpace, Type type) {
...
+            }
+            if (isAscending) {
+                return -1;
+            }
+            return 1;
+        }
+        if (TypeTags.isStringTypeTag(lhsTypeTag) && TypeTags.isStringTypeTag(rhsTypeTag))  {
+            return codePointCompare(lhsValue.toString(), rhsValue.toString());
+        }
+        if (lhsTypeTag == TypeTags.BOOLEAN_TAG && rhsTypeTag == TypeTags.BOOLEAN_TAG)  {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:3110]: We can simple of code as below

if(lhsTypeTag == rhsTypeTag) {
swtch (lhsTypeTag) {
case BOOLEAN;
// bolean logic
case INT;
.... <other tags)
}

}

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 8]
```diff
@@ -3008,6 +2976,251 @@ private static boolean containsType(Set<Object> valueSpace, Type type) {
...
+            return Long.compare((long) lhsValue, (long) rhsValue);
+        }
+        if (lhsTypeTag == TypeTags.BYTE_TAG && rhsTypeTag == TypeTags.BYTE_TAG) {
+            return Integer.compare((int) lhsValue, (int) rhsValue);
+        }
+        if (lhsTypeTag == TypeTags.FLOAT_TAG && rhsTypeTag == TypeTags.FLOAT_TAG) {
+            // Compare(x, y) is UN if x or y or both is NaN
+            // NaN should be placed last or one before the last when () is present for
+            // CompareA(x, y) and CompareD(x, y).
+            if (Double.isNaN((double) lhsValue)) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:3123]: Shall we move these to seperate method. Compare float values. Better to move these from Typechecker

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 9]
```diff
@@ -3008,6 +2976,251 @@ private static boolean containsType(Set<Object> valueSpace, Type type) {
...
+
+    public static int compareValues(Object lhsValue, Object rhsValue, String direction) {
+        int lhsTypeTag = TypeChecker.getType(lhsValue).getTag();
+        int rhsTypeTag = TypeChecker.getType(rhsValue).getTag();
+        boolean inRelationalExpr = false;
+        if (direction.isEmpty()) {
+            inRelationalExpr = true;
+        }
+        boolean isAscending = direction.equals("ascending");
+        // Compare(x,()) is UN if x is not ().
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:3084]: Shall we add a comment difference of 0, 1 and -1 return values

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java (hunk 6 above)
Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java]: comments addressed via 3515735 & [`a648baa`](https://github.com/ballerina-platform/ballerina-lang/pull/29064/commits/a648baa131843d3f1dcd88cb12640c6acab3e801)

**Pull request merged**